Cython/Compiler/PyrexTypes.py

```diff
@@ -3776,32 +3776,57 @@ def can_coerce_from_pyobject(self, env):
             return True
         return False
 
+    def _get_conversion_context_cls_cname(self, cname_format_str):
+        # cname_format_str is passed "prefix", "format", "tags"
+        X = "XYZABC"
+        tags = []
+        context = {}
+        optional_tag_strs = []
+        optional_arg_template_names = ['']  # these may be different (e.g. for unordered_map, map)
+            # hence have to be generated here rather than hard-coded in the utility_code.
+            # Start with an empty string so that the joined output starts with ','
+        for ix, T in enumerate(self.templates or []):
+            if ix < builtin_cpp_conversions[self.cname]:
+                tags.append(T.specialization_name())
+                context[X[ix]] = T
+            else:
+                optional_arg_template_names.append("%s=*" % self.template_type.templates[ix].name)
+                if not isinstance(T, TemplatePlaceholderType):
+                    tags.append(T.specialization_name())
+                    optional_tag_strs.append(X[ix])
+                    context[X[ix]] = T
```

Comment:
I was thinking more of something like this, to make it clearer that we're really doing two separate things in two separate loops here:

```suggestion
        if self.templates:
            templates = list(zip(self.templates, X, self.template_type.templates))
            builtin_template_count = builtin_cpp_conversions[self.cname]

            for T, template_name, _ in templates[:builtin_template_count]:
                tags.append(T.specialization_name())
                context[template_name] = T

            for T, template_name, type_template in templates[builtin_template_count:]:
                optional_arg_template_names.append("%s=*" % type_template.name)
                if not isinstance(T, TemplatePlaceholderType):
                    tags.append(T.specialization_name())
                    optional_tag_strs.append(template_name)
                    context[template_name] = T
```
(Not sure if `type_template` is a good name… not even sure what it is, right now.)